functions: reframe aggregate functions in terms of scalar functions

avikivity opened the pull request:
Currently, aggregate functions are implemented in a statefull manner.
The accumulator is stored internally in an aggregate_function::aggregate,
requiring each query to instantiate new instances (see
aggregate_function_selector's constructor, and note how it's called
from selector::new_instance()).
    
This makes aggregates hard to use in expressions, since expressions
are stateless (with state only provided to evaluate()). To facilitate
migration towards stateless expressions, we define a
stateless_aggregate_function (modeled after user-defined aggregates,
which are already stateless). This new struct defines the aggregate
in terms of three scalar functions: one to aggregate a new input into
an accumulator (provided in the first parameter), one to finalize an
accumulator into a result, and one to reduce two accumulators for
parallelized aggregation.
    
All existing native aggregate functions are converted to the new model, and
the old interface is removed. This series does not yet convert selectors to
expressions, but it does remove one of the obstacles.

Performance evaluation: I created a table with a million ints on a single-node cluster, and ran the avg() function on them. I measured the number of instructions executed with `perf stat -p $(pgrep scylla) -e instructions` while the query was running. The query executed from cache, memtables were flushed beforehand. The instruction count per row increased from roughly 49k to roughly 52k, indicating 3k extra instructions per row. While 3k instructions to execute a function is huge, it is currently dwarfed by other overhead (and will be even less important in a cluster where it CL>1 will cause non-coordinator code to run multiple times).

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4797/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4823/

avikivity commented:
Ready to review. I will follow up with performance evaluation (I expect a non-trivial drop).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/38/

avikivity commented:
Added a performance evaluation section. I think it's a reasonable degradation for the simplification the series unlocks, considering aggregates are far from our mainstream workload and typically execute using idle cycles.

avikivity commented:
@nyh @cvybhu @wmitros review ping

cvybhu reviewed (changes requested):
Looks good :+1:

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -93,6 +93,36 @@ class internal_scalar_function : public scalar_function {
+            return func(value_cast<std::tuple_element_t<Indexes, args_tuple_type>>(
+                         data_type_for<std::tuple_element_t<Indexes, args_tuple_type>>()->deserialize_value(*args[Indexes]))...);
+        }, std::index_sequence_for<Args...>());
+        return data_value(std::move(ret)).serialize_nonnull();
```
Not for functions managed by wrap_function_nonnull(), since it deals with with C++ types (not wrapped with std::optional).

Some aggregate functions can return NULL (like max()), but they aren't implemented with this helper.

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -945,7 +899,25 @@ std::ostream& user_aggregate::describe(std::ostream& os) const {
+            .result_type = long_type,
+            .argument_types = {},
+            .initial_state = data_value(int64_t(0)).serialize(),
+            .aggregation_function = make_internal_scalar_function("count_step", return_any_nonnull, +[] (int64_t accumulator) {
+                return accumulator + 1;
+            }),
```
std::function doesn't work, since we don't know the signature at the callee. I'll add some helper to remove the +.

cvybhu replied:
Something like this works and doesn't require the `+` trickery:
```diff
--- a/cql3/functions/aggregate_fcts.cc
+++ b/cql3/functions/aggregate_fcts.cc
@@ -125,7 +125,7 @@ wrap_function_autonull(null_handler nullhandler, Ret (*func)(Args...)) {
 
 template <typename Ret, typename... Args>
 shared_ptr<scalar_function>
-make_internal_scalar_function(sstring name, null_handler nullhandler, Ret (*func)(Args...)) {
+make_internal_scalar_function(sstring name, null_handler nullhandler, std::function<Ret(Args...)> func) {
     return ::make_shared<internal_scalar_function>(
             std::move(name),
             data_type_for<Ret>(),
```

cvybhu replied:
Or maybe not, you're right

avikivity replied:
Did you test it? I don't see how the compiler can deduce Ret and Args.

cvybhu replied:
I tested it but didn't remove the overload to `make_internal_scalar_function` and I though it was working x.x

It's a shame that there's no clean way to convert a lambda to an `std::function`.

avikivity replied:
It's impossible in a general way, since a class can contain multiple operator(), or operator() can be a template.

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -288,105 +288,71 @@ class impl_div_for_avg<big_decimal> {
+                        return data_type_for<Type>()->decompose(result);
+                    }),
+            .state_reduction_function = ::make_shared<internal_scalar_function>(
+                    "avg_finalizer",
```
fixed

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -428,53 +391,44 @@ static data_type uda_return_type(const ::shared_ptr<scalar_function>& ffunc, con
-    if (_initcond) {
-        os << "\n" << "INITCOND " << _sfunc->return_type()->deserialize(bytes_view(*_initcond)).to_parsable_string();
+    if (_agg.initial_state) {
+        os << "\n" << "INITCOND " << _agg.aggregation_function->return_type()->deserialize(bytes_view(*_agg.initial_state)).to_parsable_string();
```
There's an `if` that guards it. Not the patch doesn't change the code, just how we find the state.

avikivity commented:
Update:
 - added helper to avoid the awkward `+lamda`. It's still there, but in just one place and with a comment to explain it.
 - corrected the internal name of avg()'s reducer

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/311/

avikivity commented:
@nyh please review

nyh commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -123,6 +123,26 @@ wrap_function_autonull(null_handler nullhandler, Ret (*func)(Args...)) {
+shared_ptr<scalar_function>
+make_internal_scalar_function(sstring name, null_handler nullhandler, Lambda func) {
+    // "+func" decays the lambda into a pointer-to-function, so that its signature
+    // can be inferred by the other overload.
```
Wow, every time I think that maybe, just maybe, I'm done learning C++, you come up with yet another feature I never heard of ;-)

avikivity replied:
It's not really a feature, just an unexpected side effect. Unary operator + is defined for pointers, but not for class types. So C++ tries to find an implicit conversion that makes the unary operator + work, finds the implicit conversion to function pointer, and applies it.

nyh commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -931,88 +649,97 @@ aggregate_fcts::make_count_rows_function() {
 
 shared_ptr<aggregate_function>
 aggregate_fcts::make_max_dynamic_function(data_type io_type) {
```
Maybe the word "dynamic" should now be removed from this code, since it's no longer opposite to anything "static" and it will not really be clear to reader what it is trying to say.

avikivity replied:
ok

nyh commented on cql3/selection/abstract_function_selector.hh:
```diff
@@ -35,11 +35,11 @@ protected:
     static shared_ptr<factory> new_factory(shared_ptr<functions::function> fun, shared_ptr<selector_factories> factories);
 
-    abstract_function_selector(shared_ptr<functions::function> fun, std::vector<shared_ptr<selector>> arg_selectors)
+    abstract_function_selector(shared_ptr<functions::function> fun, std::vector<shared_ptr<selector>> arg_selectors, bool reserve_extra_arg = false)
```
I couldn't figure out what this new "reserve_extra_arg" parameter is supposed to do.

nyh reviewed (approved):
In general looks good, although in some respects it goes against our usual philosophy of moving from run-time OO (and you add a lot more overheads than just "OO") to compile-time template-based programming. CQL has only a finite (and small) number of built-in aggregates, so we could have kept them (like they were before) as spearate functions, and you didn't really need to change them to be OO just to keep the (different) state outside the function. But I agree it was easier (and shorter) since we already had this OO code in the user-defined aggregations to use the same approach also for builtin aggregators.

Do you have an idea of how good our existing tests in this area and all its different functions and edge cases (nulls, no rows, etc.) - and if they would have caught regressions? Did existing tests help you catch regressions you might have introduced during this work? I recently added some aggregation tests (test/cql-pytest/test_aggregate.py) but these target specific bugs I found recently and don't cover the entire feature. Do we have other tests with better coverage?

avikivity commented:
> In general looks good, although in some respects it goes against our usual philosophy of moving from run-time OO 

What do you mean by OO here? Object orientation? I reduced the depth of the inheritance hierarchy, and moved to composition instead. I wanted to reduce it even more (make UDA just another case of stateless aggregates) but that requires yet more work.

> (and you add a lot more overheads than just "OO") to compile-time template-based programming. CQL has only a finite (and small) number of built-in aggregates, so we could have kept them (like they were before) as spearate functions, and you didn't really need to change them to be OO just to keep the (different) state outside the function. But I agree it was easier (and shorter) since we already had this OO code in the user-defined aggregations to use the same approach also for builtin aggregators.

I do want to perform the opposite conversion. I figured it's easier to convert everything to use one type system (serialized data), then switch to stateless execution (next series), then allow multiple state types (native and serialized, unplanned) than to support both stateless native types and stateless serialized types in one go.

> 
> Do you have an idea of how good our existing tests in this area and all its different functions and edge cases (nulls, no rows, etc.) - and if they would have caught regressions? Did existing tests help you catch regressions you might have introduced during this work? I recently added some aggregation tests (test/cql-pytest/test_aggregate.py) but these target specific bugs I found recently and don't cover the entire feature. Do we have other tests with better coverage?

The existing tests caught bugs during my work (mostly around generating the name of the selector). I can't think of missing tests, but I'm pretty bad in the area of thinking up tests to challenge my own work, so that doesn't say much.

avikivity commented:
Update:
 - remove "dynamic" form make_{min,max}_dynamic_function
 - document reserve_extra_arg parameter in abstract_function_selector

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/354/

avikivity commented:
@nyh ping
